Migrate Interaction Tool

jacobperron opened the pull request:
Resolves #221

jacobperron committed:
Move InteractionTool source files

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Get interaction tool compiling + lint

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Fix warning

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Fix lint errors

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add InteractionTool plugin description

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add InteractionTool to default.rviz

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Minor refactor

* Use smart pointer
* const variables where possible

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Update README

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented on rviz_common/include/rviz_common/render_panel.hpp:
```diff
@@ -105,7 +105,6 @@ class RVIZ_COMMON_PUBLIC RenderPanel : public QWidget
   static const Ogre::Vector3 default_camera_pose_;
 
 // TODO(wjwwood): reenable these and pass down to rviz_rendering::RenderWindow
```
I'm not sure what passing down to the render window implies.

Martin-Idel replied:
I don't think it's a good idea to pass this down to the render window. It has nothing to do with Ogre and rendering, so I don't think this should be in `rviz_rendering`. I think this comment can just be ignored and deleted.

However, the separation of concerns between the `rviz_rendering::RenderWindow` and the `rviz_common::RenderPanel` has never been properly finished (Ogre is still leaking too much into `rviz_common`), so maybe William had more in mind here...

jacobperron replied:
@wjwwood Let me know if I can remove this TODO (or update it accordingly).

wjwwood replied:
I don't remember, but I do remember that it might be required to pass information down to ogre so you could do things like cause 3d objects to be "highlighted" when clicked on (for example). But I don't have a concrete use case. If you didn't need it for this work, then you can likely remove this todo.

jacobperron replied:
I wasn't sure how to test this other than mouse around on a blank view (ie. confirming that the fallback to move tool works). I will leave this PR open until I confirm it works with interactive markers  (#98).

jacobperron replied:
Seems to work as expected with what I've got so far for the interactive marker display (https://github.com/ros2/rviz/tree/jacob/interactive_markers).
Removed TODOs in 69ee286.

Martin-Idel commented on rviz_default_plugins/src/rviz_default_plugins/tools/interaction/interaction_tool.cpp:
```diff
@@ -0,0 +1,202 @@
+    flags = Render;
+  }
+
+  {
```
If I see this correctly, this is a block to allow the weak pointer to be locked and unlocked correctly. Why not push this to its own function? This would break up the quite large method into smaller methods and enhance expressiveness.

jacobperron replied:
I've refactored the method by pulling out some of the logic into smaller methods: 5a5337c

jacobperron committed:
Add missing include

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Refactor processMouseEvent into smaller methods

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

wjwwood commented on rviz_common/src/rviz_common/render_panel.cpp:
```diff
@@ -296,7 +296,7 @@ void RenderPanel::resizeEvent(QResizeEvent * event)
 const Ogre::Vector3 RenderPanel::default_camera_pose_ = Ogre::Vector3(999999, 999999, 999999);
 
-#if 0
+// TODO(jacobperron): Is there something to do related to the render window?
```
If you didn't need to and it behaves like it does in rviz in ROS 1, then you can remove this todo.

jacobperron replied:
69ee286

wjwwood commented on rviz_default_plugins/include/rviz_default_plugins/tools/interaction/interaction_tool.hpp:
```diff
@@ -0,0 +1,111 @@
+  }
+
+  /**
+   * @brief Check if the mouse has moved from one object to another,
```
nitpick: When porting, I replaced `@brief ...` with `/// ...` as we do elsewhere, e.g. in `rcl` and `rclcpp`. Also if you choose to leave brief or something else use `\` instead of `@`.

jacobperron replied:
69ee286

wjwwood reviewed (approved):
lgtm, though I didn't test it locally.

jacobperron committed:
Use selection type "Replace" instead of "Add"

Otherwise, interactive objects are never deselected as the mouse moves around.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Address review

Remove TODOs and change doxygen style.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Enable the interaction tool by default

This is the same as RViz for ROS 1.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add missing override keyword

Resolves Clang warnings.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7953)](http://ci.ros2.org/job/ci_linux/7953/)
* Linux-aarch64 [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3984)](https://ci.ros2.org/job/ci_linux-aarch64/3984/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6475)](http://ci.ros2.org/job/ci_osx/6475/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7800)](http://ci.ros2.org/job/ci_windows/7800/)

With cf6d449

* macOS [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6477)](https://ci.ros2.org/job/ci_osx/6477/)

jacobperron committed:
Fix cpplint

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
FIx lint test: [![Build Status](https://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6478)](https://ci.ros2.org/job/ci_osx/6478/)
